### PR TITLE
Add SegmentId and test

### PR DESCRIPTION
**Pull request by mar-celine**

### CODE REVIEW

File: src/Domain/SegmentId.hpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+/// A SegmentId labels a segment of the interval [-1,1].
+/// A SegmentId is used to (partially) identify an Element.
+/// The segment is the interval [-1 + 2*(N/D), -1 + 2*(N+1)/D]
+/// where N = index_ and D = 2 ^ refinement_level_.
+class SegmentId {
+ public:
+  /// Default constructor needed for Charm++ serialization.
+  SegmentId();
+
+  /// Construct from the refinement level and an index.
```
Comment (osheamonn) [inline src/Domain/SegmentId.hpp]: Redundant comment.

File: src/Domain/SegmentId.hpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  SegmentId(size_t refinement_level, size_t index);
+
+  /// The refinement level.
```
Comment (osheamonn) [inline src/Domain/SegmentId.hpp]: Redundant, I'm gonna stop commenting on things like this in this file

File: src/Domain/SegmentId.hpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  ASSERT(0 != refinement_level_, "Called on root refinement level!");
+  return SegmentId(refinement_level_ - 1, index_ >> 1);
+}
+
+inline SegmentId SegmentId::id_of_child(Side side) const {
+  // We cannot ASSERT on the maximum level because it's only known at runtime
+  // and only known elsewhere in the code, not by SegmentId. I.e. SegmentId is
+  // too low-level to know about this.
+  if (Side::Lower == side) {
+    return SegmentId(refinement_level_ + 1, index_ << 1);
```
Comment (osheamonn) [inline src/Domain/SegmentId.hpp]: Maybe I'm just dumb but I find it hard to parse bit manipulations quickly, I think anywhere we do something like this we should comment what the intention is.

Comment (nilsdeppe) [reply src/Domain/SegmentId.hpp]: Maybe both ;)

File: src/Domain/SegmentId.hpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+/// Inequivalence operator for SegmentId.
+bool operator!=(const SegmentId& lhs, const SegmentId& rhs);
+
+//##############################################################################
+// INLINE DEFINITIONS
+//##############################################################################
+
+inline SegmentId SegmentId::id_of_parent() const {
+  ASSERT(0 != refinement_level_, "Called on root refinement level!");
+  // left-shifting index_ by 1 gives us floor(index_/2).
```
Comment (wthrowe) [inline src/Domain/SegmentId.hpp]: right-shifting

File: src/Domain/SegmentId.hpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+inline SegmentId SegmentId::id_of_child(Side side) const {
+  // We cannot ASSERT on the maximum level because it's only known at runtime
+  // and only known elsewhere in the code, not by SegmentId. I.e. SegmentId is
+  // too low-level to know about this.
+  // The child has half as many segments as the parent, so for a particular
+  // parent segment, there is both an upper and lower child segment.
+  if (Side::Lower == side) {
+    // right-shifting index_ by 1 gives us index*2.
```
Comment (wthrowe) [inline src/Domain/SegmentId.hpp]: left-shifting

File: src/Domain/SegmentId.hpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    return SegmentId(refinement_level_ + 1, index_ << 1);
+  }
+  // right-shifting index_ by 1, followed by OR with 1, gives us index*2 +1.
```
Comment (wthrowe) [inline src/Domain/SegmentId.hpp]: left-shifting

File: tests/Unit/Domain/Test_SegmentId.cpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+TEST_CASE("Unit.Domain.SegmentId", "[Domain][Unit]") {
+  for (size_t level = 1; level < 5; ++level) {
+    const double segment_length = 2.0 / two_to_the(level);
+    double midpoint = -1.0 + 0.5 * segment_length;
+    for (size_t segment_index = 0; segment_index < two_to_the(level);
+         ++segment_index) {
+      SegmentId id(level, segment_index);
+      CHECK(id.midpoint() == midpoint);
+      midpoint += segment_length;
+      CHECK(id.refinement_level() ==
```
Comment (wthrowe) [inline tests/Unit/Domain/Test_SegmentId.cpp]: All these checks are duplicating the equality operator logic.  How about testing the equality operator and then using it in the later checks?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/Domain/Test_SegmentId.cpp [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      } else {
+        CHECK(id.index() == id.id_of_parent().id_of_child(Side::Upper).index());
+        CHECK(id.refinement_level() ==
+              id.id_of_parent().id_of_child(Side::Upper).refinement_level());
+      }
+    }
+  }
+
+  // Test retrieval functions:
+  SegmentId level_2_index_3(2ul, 3ul);
```
Comment (wthrowe) [inline tests/Unit/Domain/Test_SegmentId.cpp]: No `ul`s.

File: src/Domain/SegmentId.hpp [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+/// \ingroup ComputationalDomain
+/// A SegmentId labels a segment of the interval [-1,1].
+/// A SegmentId is used to (partially) identify an Element.
+/// The segment is the interval [-1 + 2*(N/D), -1 + 2*(N+1)/D]
+/// where N = index_ and D = 2 ^ refinement_level_.
+class SegmentId {
+ public:
+  /// Default constructor needed for Charm++ serialization.
+  SegmentId();
```
Comment (nilsdeppe) [inline src/Domain/SegmentId.hpp]: make this `= default` and move the default values set in the constructor into the class definition

File: tests/Unit/Domain/Test_SegmentId.cpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+#include <catch.hpp>
+
+#include "Domain/SegmentId.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+TEST_CASE("Unit.Domain.SegmentId", "[Domain][Unit]") {
+  // Test default constructor:
+  SegmentId test_id;
```
Comment (nilsdeppe) [inline tests/Unit/Domain/Test_SegmentId.cpp:12]: Should check the default values that are set if testing this.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Domain/SegmentId.hpp [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  /// The logical coordinate of the endpoint of the segment on the given Side.
+  double endpoint(Side side) const;
+
+  /// The logical coordinate of the midpoint of the segment
+  double midpoint() const {
+    return -1.0 + (1.0 + 2.0 * index_) / two_to_the(refinement_level_);
+  }
+
+  /// Serialization for Charm++
+  void pup(PUP::er& p);  // NOLINT
```
Comment (nilsdeppe) [inline src/Domain/SegmentId.hpp:49]: should check that `pup` works correctly. See `TestHelpers.hpp`, which has a `serialize_and_deserialize` function.

File: src/Domain/SegmentId.cpp [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+void SegmentId::pup(PUP::er& p) {
+  p | refinement_level_;
+  p | index_;
+}
+
+std::ostream& operator<<(std::ostream& os, const SegmentId& id) {
+  os << 'L' << id.refinement_level() << 'I' << id.index();
+  return os;
+}
+
```
Comment (nilsdeppe) [inline src/Domain/SegmentId.cpp:25]: Wrap this function in `// LCOV_EXCL_START` and `// LCOV_EXCL_STOP` since we don't want to have to worry about it during testing.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]
